src/Core/Grid/Query/CatalogPriceRuleQueryBuilder.php

```diff
@@ -0,0 +1,210 @@
+        array $contextShopIds,
+        $contextIdLang
+    ) {
+        parent::__construct($connection, $dbPrefix);
+
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+        $this->contextShopIds = $contextShopIds;
+        $this->contextIdLang = $contextIdLang;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+
+        $qb->select(
+            'pr.id_specific_price_rule,
+            pr.name,
+            pr.from_quantity,
+            pr.reduction,
+            pr.reduction_type,
+            pr.from,
+            pr.to,
+            pr_shop.name shop,
+            pr_currency.name currency,
+            pr_country.name country,
+            pr_group.name group_name'
+        );
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters())
+            ->select('COUNT(DISTINCT pr.`id_specific_price_rule`)')
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * Gets query builder with the common sql for catalog price rule listing.
+     *
+     * @param array $filters
+     *
+     * @return QueryBuilder
+     */
+    private function getQueryBuilder(array $filters)
+    {
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->from($this->dbPrefix . 'specific_price_rule', 'pr')
+            ->leftJoin(
+                'pr',
+                $this->dbPrefix . 'shop',
+                'pr_shop',
+                'pr_shop.`id_shop` = pr.`id_shop` AND pr.`id_shop` IN (:contextShopIds)'
+            )
+            ->leftJoin(
+                'pr',
+                $this->dbPrefix . 'currency_lang',
+                'pr_currency',
+                'pr_currency.`id_currency` = pr.`id_currency` AND pr_currency.`id_lang` = :contextLangId'
+            )
+            ->leftJoin(
+                'pr',
+                $this->dbPrefix . 'country_lang',
+                'pr_country',
+                'pr_country.`id_country` = pr.`id_country` AND pr_country.`id_lang` = :contextLangId'
+            )
+            ->leftJoin(
+                'pr',
+                $this->dbPrefix . 'group_lang',
+                'pr_group',
+                'pr_group.`id_group` = pr.`id_group` AND pr_group.`id_lang` = :contextLangId'
+            )
+        ;
+
+        $this->applyFilters($qb, $filters);
+        $qb->setParameter('contextLangId', $this->contextIdLang);
+        $qb->setParameter('contextShopIds', $this->contextShopIds, Connection::PARAM_INT_ARRAY);
```

Comment:
I'm not sure but this `Connection::PARAM_INT_ARRAY` might not work with all SQL implementations 🤔 can you check ? We target mysql 5.7 to mysql 8

It's not clear but I remember having issues with this. Maybe with sqlite though (in this case, we dont care).

Reply:
Our current Doctrine/DBAL version declares that it supports mysql 5.0 and above, so i guess it's not a problem